BUG: argmin/max of object dtype with NULLs, fixes #6032

jaimefrio opened the pull request.

jaimefrio committed:
BUG: argmin/max of object dtype with NULLs, fixes #6032

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -3105,22 +3105,24 @@ OBJECT_argmax(PyObject **ip, npy_intp n, npy_intp *max_ind,
     }
-    for (; i < n; i++) {
-        ip++;
+    if (mp != NULL) {
```
This looks equivalent to `i == n`.

jaimefrio replied:
They way figured, if all entries are `NULL`, then at the end `i == n` and `mp == NULL`, but if the last entry is the first non-`NULL`, then at the end `i == n` but `mp != NULL`. It is probably a good idea to add test cases with both these situations.

charris commented on numpy/core/src/multiarray/arraytypes.c.src:
```diff
@@ -3105,22 +3105,24 @@ OBJECT_argmax(PyObject **ip, npy_intp n, npy_intp *max_ind,
     *max_ind = 0;
-    i = 1;
-    while (i < n && mp == NULL) {
+    for (i = 1; i < n && mp == NULL; ++i) {
```
I think things will be a bit simpler if the loop looks like
```
for (i = 1; i < n && ip[i] == NULL; ++i);
```
`i` will be the index of the first non-NULL on exit. At that point you can set the values for all null and return, Or set `mp` and `max_ind` to continue. 

The original looks like a mutant cross between two species of loop ;)

jaimefrio commented:
I just realized we are happily assuming that the object comparison will never fail, which we know is not always the case, especially in Python 3. Added a check to exit on first failure. And I also followed your suggestion for skipping over the NULLs, although I simplified it further by not special casing the first item.

charris commented:
LGTM. I'd probably have bailed early on all NULL, but there is always more than one way ;) Is there any error return from these function that might be appropriate for a comparison error? Oh, and IIRC, richcompare is available in all supported Python versions and falls back to cmp in Python 2 if only cmp is available. I wonder if we should stop using the fallback compare? Might be something we need to pass by the list, but it might also provide more consistent behavior for Python 2 and 3.

jaimefrio commented:
Yup, 2.6 [has it](https://docs.python.org/2.6/c-api/object.html#PyObject_RichCompareBool), and so does my latest commit.

charris commented:
Should probably mention in release notes under `Compatibility Notes` just in case this changes behavior for somebody.

charris commented:
Thanks Jaime. I'm going to update the 1.11 notes anyway and will mention this.
